cove/lib/common.py

```diff
@@ -234,3 +234,98 @@ def get_schema_validation_errors(json_data, schema_url, schema_name, current_app
         unique_validator_key = [validator_type, message, path_no_number]
         validation_errors[json.dumps(unique_validator_key)].append(value)
     return dict(validation_errors)
+
+
+def _get_schema_deprecated_paths(schema_name, schema_url, obj=None, current_path=(), deprecated_paths=None):
+    '''Get a list of deprecated paths (as tuples) in a schema'''
+    if deprecated_paths is None:
+        deprecated_paths = []
+
+    if schema_url and schema_url.startswith("http"):
+        obj = jsonref.loads(requests.get(schema_url + schema_name).text)
+    elif schema_name:
+        with open(schema_url + schema_name) as schema_file:
+            obj = jsonref.load(schema_file)
+
+    for prop, value in obj['properties'].items():
+        if current_path:
+            path = current_path + (prop,)
+        else:
+            path = (prop,)
+
+        if "deprecated" in value and path not in deprecated_paths:
+                deprecated_paths.append(path)
+
+        if value.get('type') == 'object':
+            _get_schema_deprecated_paths('', '', value, path, deprecated_paths)
+        elif value.get('type') == 'array' and value['items'].get('properties'):
+            _get_schema_deprecated_paths('', '', value['items'], path, deprecated_paths)
+
+    return deprecated_paths
+
+
+def _get_json_data_generic_paths(json_data, path=(), generic_paths=None):
+    '''Transform json data into a dictionary with keys made of json paths.
+
+   Key are json paths (as tuples). Values are dictionaries with keys including specific
+   indexes (which are not including in the top level keys), eg:
+
+   {'a': 'I am', 'b': ['a', 'list'], 'c': [{'ca': 'ca1'}, {'ca': 'ca2'}, {'cb': 'cb'}]}
+
+   will produce:
+
+   {('a',): {('a',): I am'},
+    ('b',): {{(b, 0), 'a'}, {('b', 1): 'list'}},
+    ('c', 'ca'): {('c', 0, 'ca'): 'ca1', ('c', 1, 'ca'): 'ca2'},
+    ('c', 'cb'): {('c', 1, 'ca'): 'cb'}}
+    '''
+    if generic_paths is None:
+        generic_paths = {}
+
+    if isinstance(json_data, dict):
+        iterable = list(json_data.items())
+        if not iterable:
+            generic_paths[path] = {}
+    else:
+        iterable = list(enumerate(json_data))
+        if not iterable:
+            generic_paths[path] = []
+
+    for key, value in iterable:
+        if isinstance(value, (dict, list)):
+            _get_json_data_generic_paths(value, path + (key,), generic_paths)
+        else:
+            generic_key = tuple(i for i in path + (key,) if type(i) != int)
+            if generic_paths.get(generic_key):
+                generic_paths[generic_key][path + (key,)] = value
+            else:
+                generic_paths[generic_key] = {path + (key,): value}
+
+    return generic_paths
+
+
+def get_json_data_deprecated_fields(schema_name, schema_url, json_data):
+    deprecated_schema_paths = _get_schema_deprecated_paths(schema_name, schema_url)
+    paths_in_data = _get_json_data_generic_paths(json_data)
```

Comment:
I'd be interested to know how much memory this uses, particularly since this gets called even if their are no deprecated paths in the schema.